pipeline_tools/http_requests.py

```diff
@@ -0,0 +1,272 @@
+        self.record_dir = self._get_param(HTTP_RECORD_DIR, '.')
+        self.retry_timeout = self._get_param(RETRY_TIMEOUT, '600', int)
+        self.retry_max_tries = self._get_param(RETRY_MAX_TRIES, '10000', int) 
+        self.retry_multiplier = self._get_param(RETRY_MULTIPLIER, '1', float)
+        self.retry_max_interval = self._get_param(RETRY_MAX_INTERVAL, '10', float)
+
+    def _get_param(self, name, default, convert_fn=None):
+        param_str = os.environ.get(name, default)
+        # handle case where environment var is set to empty string
+        if not param_str:
+            param_str = default
+
+        if convert_fn is not None:
+            return convert_fn(param_str)
+        else:
+            return param_str
+
+    def get(self, *args, **kwargs):
+        """Calls requests.get function.
+        
+        In addition to calling requests.get, this function will record the request
+        and response if the HttpRequests object's should_record attribute is True.
+
+        Args:
+            All arguments that requests.get accepts are accepted here.
+
+        Returns:
+            A requests.Response object representing the response to the request.
+
+        Raises:
+            ValueError if response status code indicates an error.
+        """
+        kwargs['method'] = 'get'
+        return self._http_request_with_retry(*args, **kwargs)
+
+    def put(self, *args, **kwargs):
+        """Calls requests.put function.
+        
+        In addition to calling requests.put, this function will record the request
+        and response if the HttpRequests object's should_record attribute is True.
+
+        Args:
+            All arguments that requests.put accepts are accepted here except
+            for 'body' -- use 'json' instead.
+
+        Returns:
+            A requests.Response object representing the response to the request.
+
+        Raises:
+            ValueError if response status code indicates an error.
+        """
+        kwargs['method'] = 'put'
+        return self._http_request_with_retry(*args, **kwargs)
+
+    def post(self, *args, **kwargs):
+        """Calls requests.post function.
+        
+        In addition to calling requests.post, this function will record the request
+        and response if the HttpRequests object's should_record attribute is True.
+
+        Args:
+            All arguments that requests.post accepts are accepted here except
+            for 'body' -- use 'json' instead.
+
+        Returns:
+            A requests.Response object representing the response to the request.
+
+        Raises:
+            ValueError if response status code indicates an error.
+        """
+        kwargs['method'] = 'post'
+        return self._http_request_with_retry(*args, **kwargs)
+
+    def _http_request_with_retry(self, *args, **kwargs):
+        def _is_retryable_status_code(error):
+            if not isinstance(error, requests.HTTPError):
```

Comment:
Could we simplify this to `return isinstance(error, requests.HTTPError) and not (400 <= error.response.status_code <= 499)`? I find checks like this easier to understand when there are no double-negatives!